src/docfx/lib/git/FileCommitProvider.cs

```diff
@@ -0,0 +1,329 @@
+                        break;
+                    }
+                }
+                if (!found)
+                {
+                    continue;
+                }
+
+                // Lookup and use cached commit history ONLY if there are no other commits to follow
+                if (commitsToFollow.Count == 0)
+                {
+                    lock (commitCache)
+                    {
+                        if (commitCache.TryGetValue((commit.Sha.a, blob), out var cachedValue))
+                        {
+                            updateCache = result.Count != 0;
+
+                            var (cachedCommitHistory, lruOrder) = cachedValue;
+                            foreach (var cachedCommit in cachedCommitHistory)
+                            {
+                                result.Add(commitsBySha[cachedCommit]);
+                            }
+                            commitCache[(commit.Sha.a, blob)] = (cachedCommitHistory, _nextLruOrder--);
+                            break;
+                        }
+                    }
+                }
+
+                var singleParent = false;
+                var parentCount = Math.Min(MaxParentBlob, commit.Parents.Length);
+                var add = parentCount == 0 && blob != 0;
+
+                for (var i = 0; i < parentCount; i++)
+                {
+                    parentBlobs[i] = GetBlob(commit.Parents[i].Tree, pathSegments);
+                    if (parentBlobs[i] == blob)
+                    {
+                        // and it was TREESAME to one parent, follow only that parent.
+                        // (Even if there are several TREESAME parents, follow only one of them.)
+                        commitsToFollow.Add((commit.Parents[i], blob));
+                        singleParent = true;
+                        break;
+                    }
+                }
+
+                if (!singleParent)
+                {
+                    // Otherwise, follow all parents.
+                    for (var i = 0; i < parentCount; i++)
+                    {
+                        add = true;
+                        commitsToFollow.Add((commit.Parents[i], parentBlobs[i]));
+                    }
+                }
+
+                if (add)
+                {
+                    result.Add(commit);
+                }
+            }
+
+            if (updateCache)
+            {
+                lock (commitCache)
+                {
+                    _cacheUpdated = true;
+                    commitCache.Add((headCommit.Sha.a, headBlob), (result.Select(c => c.Sha.a).ToArray(), 0));
+                }
+            }
+
+            return result.Select(c => c.GitCommit).ToList();
+        }
+
+        public IReadOnlyList<KeyValuePair<string, Dictionary<(long commit, long blob), (long[] commitHistory, int lruOrder)>>> BuildCache()
+        {
+            if (_cacheUpdated)
+            {
+                return _commitCache.ToArray();
```

Comment:
nit: Do you want to add a comment about why it is changed from `ToList` to `ToArray`?

Reply:
ok, will add it

---
In reply to: [237942785](https://github.com/dotnet/docfx/pull/3705#discussion_r237942785) [](ancestors = 237942785)